### PR TITLE
Infer importpath if not set explicitly

### PR DESCRIPTION
**Pull request by mering**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

This infers the `importpath` automatically if it is not set. If set explicitly it continues to use this one.

**Which issues(s) does this PR fix?**

**Other notes for review**

### CODE REVIEW

File: go/private/context.bzl [hunk 1]
```diff
@@ -357,17 +357,19 @@ def _check_importpaths(ctx):
         if ":" in p:
             fail("import path '%s' contains invalid character :" % p)
 
-def _infer_importpath(ctx):
+def _infer_importpath(ctx, attr = None):
```
Comment (fmeum) [inline go/private/context.bzl:360]: I think that this is only called in one place and that call site provides `attr`, so we can make this attribute required.

**Review by fmeum — APPROVED** [review]
LGTM, thanks!

@linzhp Do you have any concerns?

File: go/private/context.bzl (hunk 1 above)
Comment (mering) [reply go/private/context.bzl]: Good idea. Done.

**Review by linzhp — APPROVED** [review]
LGTM

File: tests/core/go_proto_library_importpath/bar.proto [hunk 2]
```diff
@@ -0,0 +1,9 @@
+syntax = "proto3";
+
+package tests.core.go_proto_library_importpath.bar;
```
Comment (linzhp) [inline tests/core/go_proto_library_importpath/bar.proto:3]: Actually, the mapping from "tests.core.go_proto_library_importpath.bar" to "tests/core/go_proto_library_importpath/bar_go_proto" sounds very opinionated. Can we only infer the import path when `go_package` is set?

Comment (mering) [reply tests/core/go_proto_library_importpath/bar.proto:3]: The logic to infer the importpath is defined in https://github.com/bazelbuild/rules_go/blob/69d0fc88d0815bd11867ed3f26f36244faa3c06c/go/private/context.bzl#L384-L393 which I do not touch in this PR. It is basically `package/label` which seems logical within the Bazel world. This is also what we are currently using internally via a macro on top of the `go_proto_library` rule. To me, it sounds reasonable to keep the current behavior while at the same time maintaining the possibility of overwriting the importpath as before. So for existing users for which the explicit setting of importpath was obligatory, the behavior doesn't change as well.

Comment (linzhp) [reply tests/core/go_proto_library_importpath/bar.proto:3]: I see. Thanks for the explanation.

**Pull request merged**